internal/xds/bootstrap/bootstrap.go

```diff
@@ -141,38 +213,38 @@ func (sc *ServerConfig) CredsDialOption() grpc.DialOption {
 // content. It doesn't cover NodeProto because NodeProto isn't used by
 // federation.
 func (sc *ServerConfig) String() string {
-	features := strings.Join(sc.ServerFeatures, "-")
-	return strings.Join([]string{sc.ServerURI, sc.Creds.String(), features}, "-")
+	features := strings.Join(sc.serverFeatures, "-")
+	return strings.Join([]string{sc.serverURI, sc.selectedCreds.String(), features}, "-")
 }
 
-// MarshalJSON marshals the ServerConfig to json.
-func (sc ServerConfig) MarshalJSON() ([]byte, error) {
-	server := xdsServer{
-		ServerURI:      sc.ServerURI,
-		ChannelCreds:   []channelCreds{{Type: sc.Creds.Type, Config: sc.Creds.Config}},
-		ServerFeatures: sc.ServerFeatures,
-	}
-	server.ServerFeatures = []string{serverFeaturesV3}
-	if sc.IgnoreResourceDeletion {
-		server.ServerFeatures = append(server.ServerFeatures, serverFeaturesIgnoreResourceDeletion)
+// The following fields correspond 1:1 with the JSON schema for ServerConfig.
+type serverConfigJSON struct {
+	ServerURI      string         `json:"server_uri"`
+	ChannelCreds   []ChannelCreds `json:"channel_creds"`
+	ServerFeatures []string       `json:"server_features"`
+}
+
+// MarshalJSON returns marshaled JSON bytes corresponding to this server config.
+func (sc *ServerConfig) MarshalJSON() ([]byte, error) {
+	server := &serverConfigJSON{
+		ServerURI:      sc.serverURI,
+		ChannelCreds:   sc.channelCreds,
+		ServerFeatures: sc.serverFeatures,
 	}
 	return json.Marshal(server)
 }
 
 // UnmarshalJSON takes the json data (a server) and unmarshals it to the struct.
 func (sc *ServerConfig) UnmarshalJSON(data []byte) error {
-	var server xdsServer
+	server := serverConfigJSON{}
 	if err := json.Unmarshal(data, &server); err != nil {
-		return fmt.Errorf("xds: json.Unmarshal(data) for field ServerConfig failed during bootstrap: %v", err)
+		return fmt.Errorf("xds: json.Unmarshal(%s) for server configuration failed during bootstrap: %v", string(data), err)
```

Comment:
> the json package are not very readable
you are very right about that! It sometimes refers to the char count which is totally unreadable. 

I'm happy with how it looks now. Otherwise i think it's still unreadable trying to parse the `json.Unmarshal`'s `()`.